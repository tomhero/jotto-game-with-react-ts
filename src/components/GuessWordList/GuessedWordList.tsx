import React from 'react'

export type guessedWord = {
  guessedWord: string;
  letterMatchCount: number;
};

export type GuessedWordListProps = {
  guessedWords: guessedWord[]
};

const GuessedWordList: React.FC<GuessedWordListProps> = (props) => {
  const guessedWordRows = props.guessedWords.map((word, index) => {
    return (
      <tr key={`guessed-word-${index}`} data-test="guessed-word">
        <td>{word.guessedWord}</td>
        <td>{word.letterMatchCount}</td>
      </tr>
    );
  });
  return (
    <div data-test="component-guessed-word-list">
      {props.guessedWords.length === 0 ?
        (
          <h2 data-test="guess-instructions">
            Plese guess a word
          </h2>
        ) : (
          <section data-test="guessed-word-list-section">
            <h3>Guessed Words</h3>
            <table>
              <thead>
                <tr>
                  <th>Matching Letters</th>
                </tr>
              </thead>
              <tbody>
                {guessedWordRows}
              </tbody>
            </table>
          </section>
        )
      }
    </div>
  )
}

export default GuessedWordList;
