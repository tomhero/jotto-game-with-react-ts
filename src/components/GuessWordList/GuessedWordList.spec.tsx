import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import GuessedWordList, { guessedWord, GuessedWordListProps } from './GuessedWordList';
import { findByTesttAttr } from '../../../test/test-utils';

describe('<GuessedWordList />', () => {
  let guessedWordListWrapper: ShallowWrapper;
  let defaultProps: GuessedWordListProps = {
    guessedWords: [],
  };

  beforeEach(() => {
    guessedWordListWrapper = shallow(<GuessedWordList {...defaultProps} />);
  });

  describe('if there are no words guessed', () => {
    beforeEach(() => {
      guessedWordListWrapper.setProps({ guessedWords: [] });
    });

    it('renders the component without error', () => {
      expect(guessedWordListWrapper).toMatchSnapshot();
      const containerComponent = findByTesttAttr(guessedWordListWrapper, 'component-guessed-word-list');
      expect(containerComponent).toHaveLength(1);
    });

    it('renders instructions to guess a word', () => {
      const instructions = findByTesttAttr(guessedWordListWrapper, 'guess-instructions');
      expect(instructions.text()).not.toBe('');
    });
  });

  describe('if there are words guessed', () => {
    const mockingWordList: guessedWord[] = [
        { guessedWord: 'qwerty', letterMatchCount: 3 },
        { guessedWord: 'asdfgh', letterMatchCount: 1 },
        { guessedWord: 'inquiry', letterMatchCount: 7 },
    ];
    beforeEach(() => {
      guessedWordListWrapper.setProps({
        guessedWords: mockingWordList,
      });
    });

    it('renders the component without error', () => {
      expect(guessedWordListWrapper).toMatchSnapshot();
      const containerComponent = findByTesttAttr(guessedWordListWrapper, 'component-guessed-word-list');
      expect(containerComponent).toHaveLength(1);
    });

    it('renders guessed word list table', () => {
      const wordList = findByTesttAttr(guessedWordListWrapper, 'guessed-word-list-section');
      expect(wordList).toHaveLength(1);
    });

    it('has correct number of guessed words', () => {
      const guessedWordNodes = findByTesttAttr(guessedWordListWrapper, 'guessed-word');
      expect(guessedWordNodes).toHaveLength(mockingWordList.length);
    });
  });
});

