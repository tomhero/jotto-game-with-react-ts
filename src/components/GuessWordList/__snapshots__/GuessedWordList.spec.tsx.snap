// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GuessedWordList /> if there are no words guessed renders the component without error 1`] = `
<div
  data-test="component-guessed-word-list"
>
  <h2
    data-test="guess-instructions"
  >
    Plese guess a word
  </h2>
</div>
`;

exports[`<GuessedWordList /> if there are words guessed renders the component without error 1`] = `
<div
  data-test="component-guessed-word-list"
>
  <section
    data-test="guessed-word-list-section"
  >
    <h3>
      Guessed Words
    </h3>
    <table>
      <thead>
        <tr>
          <th>
            Matching Letters
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          data-test="guessed-word"
          key="guessed-word-0"
        >
          <td>
            qwerty
          </td>
          <td>
            3
          </td>
        </tr>
        <tr
          data-test="guessed-word"
          key="guessed-word-1"
        >
          <td>
            asdfgh
          </td>
          <td>
            1
          </td>
        </tr>
        <tr
          data-test="guessed-word"
          key="guessed-word-2"
        >
          <td>
            inquiry
          </td>
          <td>
            7
          </td>
        </tr>
      </tbody>
    </table>
  </section>
</div>
`;
